<!-- R Markdown Documentation, DO NOT EDIT THE PLAIN MARKDOWN VERSION OF THIS FILE -->

<!-- Copyright 2015 Google Inc. All rights reserved. -->

<!-- Licensed under the Apache License, Version 2.0 (the "License"); -->
<!-- you may not use this file except in compliance with the License. -->
<!-- You may obtain a copy of the License at -->

<!--     http://www.apache.org/licenses/LICENSE-2.0 -->

<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# ExAC Variant Analysis

What follows is a brief exploration of the public data from [ExAC Release 0.2](ftp://ftp.broadinstitute.org/pub/ExAC_release/release0.2/README.release0.2).  

_Have feedback or corrections?  Please submit a pull request or file an issue!_

```{r echo=FALSE, eval=FALSE}
######################[ CHANGE ME ]##################################
# This codelab assumes that the current working directory is where the Rmd file resides.
setwd("/YOUR/PATH/TO/codelabs/R/ExAC-analysis")

# Set the Google Cloud Platform project id under which these queries will run.
project <- "YOUR-PROJECT-ID"
#####################################################################
```

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE}
# Set up for BigQuery access.
source("../PlatinumGenomes-QC/rHelpers/setup.R")
```

* [Multi-Allelic Variants by Chromosome](#multi-allelic-variants-by-chromosome)
 * [Normalized Multi-Allelic Variants by Chromosome](#normalized-multi-allelic-variants-by-chromosome)
* [Shared Variation](#shared-variation)
* [Shared Variation By Population](#shared-variation-by-population)
 * [All populations plotted together](#all-populations-plotted-together)
 * [Populations plotted individually](#populations-plotted-individually)
 * [Populations plotted individually overlaid by the variation among all](#populations-plotted-individually-overlaid-by-the-variation-among-all)
* [Burden of Variation](#burden-of-variation)
* [Provenance](#provenance)

## Multi-Allelic Variants by Chromosome

From [A Guide to the Exome Aggregation Consortium (ExAC) Data Set](http://macarthurlab.org/2014/11/18/a-guide-to-the-exome-aggregation-consortium-exac-data-set/) by the MacArthur Lab:

> There is increasing urgency for the development of tools that deal appropriately with these mult-iallelic sites - approximately ~7% of ExAC sites are now multi-allelic, and that fraction will grow as our sample size increases. That high rate of multiallelism shouldn't be surprising, by the way - the ExAC data set now (staggeringly) contains one variant every six bases on average, so it's not a shock to see many cases where variant locations overlap.

```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("./sql/multi-allelic-variants.sql",
                                  project=project)
```
Number of rows returned by this query: **`r if(is.null(result)) { "None" } else { nrow(result) }`**.

Displaying the first few rows of the dataframe of results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
if(is.null(result)) { 
  cat("**None**") 
} else {
  print(xtable(head(result)), type="html", include.rownames=F)
}
```

```{r}
result$reference_name <- factor(result$reference_name, levels=c(as.character(seq(1,22)), "X", "Y"))
result$num_alternates <- factor(result$num_alternates) 
```

Visualizing the results:
```{r multiAllelicLogScale, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(result, aes(x=reference_name, y=num_multi_allelic_variants,  fill=num_alternates)) +
  geom_bar(stat="identity") +
  xlab("Chromosome") +
  scale_y_log10(labels=comma) +
  scale_fill_discrete("Number of\nAlternate\nAlleles") +
  ylab("Number of Variants (log scale)") +
  ggtitle("Variant Count by Number of Alternate Alleles") +
  facet_grid(num_alternates ~ .)
```

### Normalized Multi-Allelic Variants by Chromosome

Let's normalize these results by [exon bases per chromosome](http://seqanswers.com/forums/showthread.php?t=5298) and take a closer look.

```{r}
basesPerChr <- read.csv("./data/bases-per-chromosome.csv")
result <- inner_join(result, basesPerChr)
result$reference_name <- factor(result$reference_name, levels=c(as.character(seq(1,22)), "X", "Y"))
```

```{r multiAllelicNormalized, fig.align="center", fig.width=10, message=FALSE, comment=NA}
for(n in levels(result$num_alternates)) {
  p <- ggplot(filter(result, num_alternates == n),
              aes(x=reference_name,
                  y=num_multi_allelic_variants/Exon.Bases,
                  fill=num_alternates)) +
    geom_bar(stat="identity") +
    scale_fill_discrete("Number of\nAlternate\nAlleles") +
    xlab("Chromosome") +
    ylab("Variant Count / Chromosome Exome Bases Count") +
    scale_y_continuous(labels=percent_format()) +
    ggtitle("Variant Count as Percentage of Exomes Bases\nby Number of Alternate Alleles")
  print(p)
}
```

After normalizing the counts by the number of exon bases per chromosome, it is 
interesting to note that chromosome 14 appears to have a larger number of high 
multi-allelic sites compared to the other chromosomes.

## Shared Variation

Let explore shared variation in a manner similar to [what we did for 1,000 Genomes](https://github.com/googlegenomics/bigquery-examples/blob/master/1000genomes/README.md#exploring-shared-variation).

ExAC contains site information from 61,486 samples.  To paraphrase some helpful clarifications from [Daniel MacArthur](http://macarthurlab.org/people/): 

* The number of samples can vary by site, as there are some parts of the exome where (due to varying coverage) only a fraction of the samples were callable. 
* This number is indicated by the AN tag at each site in the VCF, which equates to the number of callable haplotypes (i.e. 2 x the number of callable individuals).
* Often AC_Adj (the allele count at a site, counting only high-confidence genotype calls) and AN_Adj (the total number of callable haplotypes at a site, counting only high-confidence genotypes) are used as a robust estimate of allele frequency at a given site.

```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("./sql/shared-variant-counts.sql",
                                  project=project)
```
Number of rows returned by this query: **`r if(is.null(result)) { "None" } else { nrow(result) }`**.

Displaying the first few rows of the dataframe of results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
if(is.null(result)) { 
  cat("**None**") 
} else {
  print(xtable(head(result)), type="html", include.rownames=F)
}
```

```{r sharedVariants, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(result, aes(x=percentage_of_alleles_with_variant,
                   y=number_of_variants_shared_by_this_percentage_of_alleles)) +
  geom_point(alpha=1/3) +
  scale_x_continuous(labels=percent_format()) +
  scale_y_log10(labels=comma) +
  ylab("Count of variants shared by X% of the alleles (log scale)") +
  xlab("Percentage of alleles sharing a variant") +
  ggtitle("An overview of how variation is shared across alleles")
```

```{r}
# Get these results ready for later use
allResults <- mutate(result,
                     population="ALL",
                     description="All Populations"
                     )
```

## Shared Variation by Population

Let's drill down further by population.  Note that we pivot our data into a [tidy data format](http://vita.had.co.nz/papers/tidy-data.pdf) by taking advantage of the [CROSS JOIN](https://cloud.google.com/bigquery/query-reference#joins) feature of BigQuery.

```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("./sql/shared-variant-counts-by-population.sql",
                                  project=project)
```
Number of rows returned by this query: **`r if(is.null(result)) { "None" } else { nrow(result) }`**.

Displaying the first few rows of the dataframe of results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
if(is.null(result)) { 
  cat("**None**") 
} else {
  print(xtable(head(result)), type="html", include.rownames=F)
}
```

### All populations plotted together
```{r sharedVariantsAllPops, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(result, aes(x=percentage_of_alleles_with_variant,
                   y=number_of_variants_shared_by_this_percentage_of_alleles,
                   color=population)) +
  geom_point(alpha=1/3) +
  scale_x_continuous(labels=percent_format()) +
  scale_y_log10(labels=comma) +
  ylab("Count of variants shared by X% of the alleles (log scale)") +
  xlab("Percentage of alleles sharing a variant") +
  ggtitle("An overview of how variation is shared across alleles")
```

### Populations plotted individually
```{r sharedVariantsByPop, fig.align="center", fig.width=10, message=FALSE, comment=NA}
for(pop in unique(result$population)) {
  p <- ggplot(filter(result, population == pop),
              aes(x=percentage_of_alleles_with_variant,
                   y=number_of_variants_shared_by_this_percentage_of_alleles,
                   color=population)) +
  geom_point(alpha=1/3) +
  scale_x_continuous(labels=percent_format()) +
  scale_y_log10(labels=comma) +
  ylab("Count of variants shared by X% of the alleles (log scale)") +
  xlab("Percentage of alleles sharing a variant") +
  ggtitle("An overview of how variation is shared across alleles")
  print(p)
}
```

### Populations plotted individually overlaid by the variation among all
```{r sharedVariantsByPopOverlayAll, fig.align="center", fig.width=10, message=FALSE, comment=NA}
for(pop in unique(result$population)) {
  popVsAll <- rbind(allResults, filter(result, population == pop))
  popVsAll$population <- factor(popVsAll$population, levels=c("ALL", pop))
  p <- ggplot(popVsAll,
              aes(x=percentage_of_alleles_with_variant,
                   y=number_of_variants_shared_by_this_percentage_of_alleles,
                   color=population)) +
  geom_point(alpha=1/3) +
  scale_x_continuous(labels=percent_format()) +
  scale_y_log10(labels=comma) +
  ylab("Count of variants shared by X% of the alleles (log scale)") +
  xlab("Percentage of alleles sharing a variant") +
  ggtitle("An overview of how variation is shared across alleles")
  print(p)
}
```

# Burden of Variation

What follows is a sample query upon individual-level data from 1,000 Genomes phase 1 variants and SNP annotations from Tute Genomics.  Note that the [1,000 Genomes table](https://bigquery.cloud.google.com/table/genomics-public-data:1000_genomes.variants) has 39,728,277 rows and the [Tute Annotations table](https://bigquery.cloud.google.com/table/silver-wall-555:TuteTable.hg19) has 8,582,972,385 rows.

We calculate the null spectrum (burden of variation) for each gene: how many individuals have 2 or more missense mutations at < 2% frequency in gene ____?

```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("./sql/burden-of-variation.sql",
                                  project=project)
```
Number of rows returned by this query: **`r if(is.null(result)) { "None" } else { nrow(result) }`**.

Displaying the first few rows of the dataframe of results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
if(is.null(result)) { 
  cat("**None**") 
} else {
  print(xtable(head(result)), type="html", include.rownames=F)
}
```

Let's take a look at the burden of variation within BRCA1:
```{r}
brca1Result <- filter(result, grepl("BRCA1", Gene))
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
if(is.null(result)) { 
  cat("**None**") 
} else {
  print(xtable(brca1Result), type="html", include.rownames=F)
}
```

And let's also see which individuals carry the most burden of variation:
```{r}
sampleResult <- arrange(summarize(group_by(result, call_call_set_name), total = sum(number_of_rare_missense_variants)), desc(total))
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
if(is.null(result)) { 
  cat("**None**") 
} else {
  print(xtable(head(sampleResult)), type="html", include.rownames=F)
}
```

# Provenance

* The source data from ftp://ftp.broadinstitute.org/pub/ExAC_release/release0.2/ExAC.r0.2.sites.vep.vcf.gz was loaded into Google Genomics and exported to BigQuery as table [google.com:biggene:ExAC_release_0_2.variants](https://bigquery.cloud.google.com/table/google.com:biggene:ExAC_release_0_2.variants).  
* We then flattened the resulting table to make it a bit easier to query by materializing the result of query [reshape-data.sql](./sql/reshape-data.sql) to table [google.com:biggene:ExAC_release_0_2.reshaped_variants](https://bigquery.cloud.google.com/table/google.com:biggene:ExAC_release_0_2.reshaped_variants).
  * Note that the upcoming User-Defined Function feature for BigQuery will make this sort of transformation easier to write.  See [reshape-data-udf.sql](./sql/reshape-data-udf.sql) as an example.

```{r}
sessionInfo()
```
